test_dapps/packages/test_app/app/js/index.js

```diff
@@ -159,7 +159,7 @@ $(document).ready(function() {
   $("#communication button.listenToChannel").click(function() {
     var channel = $("#communication .listen input.channel").val();
     $("#communication #subscribeList").append("<br> subscribed to " + channel + " now try sending a message");
-    EmbarkJS.Messages.listenTo({topic: [channel]}).then(function(message) {
+    EmbarkJS.Messages.listenTo({topic: [channel]}).subscribe(function({data: message}) {
```

Comment:
:100: much nicer

Reply:
What's unclear to me though is why does the emitted value have to be wrapped inside `{ data }`?

This is kind of a breaking change.

Reply:
@PascalPrecht good question.

You may have already figured out some of this, but let me preface in case others weigh in:

The [implementation of listenTo](https://github.com/embark-framework/embark/blob/master/packages/embark/src/lib/modules/whisper/js/communicationFunctions.js#L92-L95) this code replaces (see also: [code supporting older web3](https://github.com/embark-framework/embark/blob/master/packages/embark/src/lib/modules/whisper/js/embarkjs_old_web3.js#L96)) is a mix of a pseudo-promise and callback. The `.cb` function on [L95](https://github.com/embark-framework/embark/blob/master/packages/embark/src/lib/modules/whisper/js/communicationFunctions.js#L95) is registered when someone uses the pseudo-promise's `.then` method.

In any case, it needed to be changed to return a promise or invoke a callback (if one is supplied) and not do both. But considering how it works — firing functions registered w/ `.then` multiple times — it's really not a promise at all. Since the pseudo-promise already behaved like an observable, I modified `listenTo` to return actual RxJS Observable instances, which have a well defined API and are a good match for this use case, i.e. subscribing to whisper channels.

When pseudo-promise's `then` function is called, it receives 3 arguments, the first of which corresponds to `message` in the line you commented on and this [L81](https://github.com/embark-framework/embark/blob/master/packages/embark/src/lib/modules/whisper/js/communicationFunctions.js#L81).  3 arguments being passed to the [`next`](https://rxjs.dev/api/index/interface/Subscribable) function of a Subscribable didn't seem to be a common pattern, so I consolidated the data object that's received by any subscriber's next method.  It's a breaking change but seemed wortwhile. I am very open to suggestions on how to make it better.

Reply:
Thanks for taking the time to elaborate on this so thoroughly - really appreciate that!

> firing functions registered w/ `.then` multiple times — it's really not a promise at all. Since the pseudo-promise already behaved like an observable, I modified `listenTo` to return RxJS Observable instances, which have a well defined API and are a good match for this use case, i.e. subscribing to whisper channels.

I fully agree that neither promises nor something that looks like a promise but behaves differently is a good fit for such an API. Moving to something like Rx for event streams makes a lot of sense to me. 👌

> Three arguments being passed to the [`next`](https://rxjs.dev/api/index/interface/Subscribable) function of a Subscribable didn't seem to be a common pattern, so I consolidated the data object that's received by a subscriber's next method. It's a breaking change but seemed worthwhile.

Yes, that makes sense. As a matter of fact, the moment we switched from fake-promise to Observables, this was already a breaking change. It doesn't matter too much anymore that the emitted value comes in `data` as any call-side of this API will need update.

**However**, either way this is a breaking change and I'd like us to make this as clear as possible to our users (even if chances are low that many people are using Embark's Messages interface at the moment.

Luckily we can put those in our commit message and will be parsed so they end up nicely in our changelogs: https://github.com/embark-framework/embark/blob/master/CONTRIBUTING.md#footer

Would you mind updating this accordingly, or do you want me to do it on your commit really quick?